### DISCUSSION TITLE
Bidirectional, polymorphic effects

### DISCUSSION THREAD

**Discussion by anfelor** [koka-lang/koka#427] [Q&A] [answered]
Hi @daanx,

Wenhao (@thwfhk) and I were thinking about encoding bidirectional effects in Koka. To do this, an effect would return a thunk which can itself perform effects. Naturally, it would be nice to allow polymorphism over the effects of the thunk, as in:

```koka
effect ask<a,e>
  fun ask() : (() -> <e> a)
```

Unfortunately, this is rejected by Koka as:

```
invalid syntax
encountered effect variable e when an effect label was expected
  hint: effect variables must be after `|` (e.g `<labels|e>`), or by themselves (e.g. `e`)
```

Does Koka support this use-case?

**Comment by TimWhiting** [comment, marked as answer]
Try doing
```
effect ask<a,e>
  fun ask() : (() -> e a)
```
Putting the e inside the angle brackets causes it to be interpreted as an effect label not a polymorphic effect variable.
